src/main/java/org/apache/sling/xss/impl/CustomPolicy.java

```diff
@@ -0,0 +1,265 @@
+    private CssValidator cssValidator;
+
+    public CustomPolicy(Policy policy) {
+        removeAttributeGuards();
+        HtmlPolicyBuilder policyBuilder = new HtmlPolicyBuilder();
+
+        cssValidator = new CssValidator(policy.getCssPolicy());
+
+        // ------------ this is for the global attributes -------------
+        Map<String, Attribute> globalAttributes = policy.getGlobalAttributes();
+        for (Attribute attribute : globalAttributes.values()) {
+
+            if (attribute.getOnInvalid().equals("removeTag")) {
+                onInvalidRemoveTagList.add(attribute.getName());
+            }
+
+            if (CssValidator.STYLE_ATTRIBUTE_NAME.equals(attribute.getName())) {
+                // we match style tags separately
+                policyBuilder.allowAttributes(attribute.getName()).matching(cssValidator.newCssAttributePolicy())
+                        .globally();
+            } else {
+                List<String> allowedValuesFromAttribute = attribute.getLiterals();
+                if (allowedValuesFromAttribute != null && allowedValuesFromAttribute.size() > 0) {
+                    for (String allowedValue : allowedValuesFromAttribute) {
+                        policyBuilder.allowAttributes(attribute.getName()).matching(true, allowedValue).globally();
+                    }
+
+                }
+                List<Pattern> regexsFromAttribute = attribute.getPatternList();
+                if (regexsFromAttribute != null && regexsFromAttribute.size() > 0) {
+                    policyBuilder.allowAttributes(attribute.getName()).matching(matchesToPatterns(regexsFromAttribute))
+                            .globally();
+                } else {
+                    policyBuilder.allowAttributes(attribute.getName()).globally();
+                }
+
+            }
+        }
+
+        // ------------ this is for the allowed emty tags -------------
+        List<String> allowedEmptyTags = policy.getAllowedEmptyTags();
+        for (String allowedEmptyTag : allowedEmptyTags) {
+            policyBuilder.allowWithoutAttributes(allowedEmptyTag);
+        }
+
+        // ------------ this is for the tag rules -------------
+        Map<String, Tag> tagMap = policy.getTagRules();
+        for (Map.Entry<String, Tag> tag : tagMap.entrySet()) {
+
+            String tagAction = tag.getValue().getAction();
+            switch (tagAction) {
+            // Tag.action
+            case "truncate":
+                policyBuilder.allowElements(tag.getValue().getName());
+
+                break;
+            // filter: remove tags, but keep content,
+            case "filter":
+                break;
+            // remove: remove tag and contents
+            case "remove":
+                policyBuilder.disallowElements(tag.getValue().getName());
+                break;
+
+            // validate is also the default
+            // validate: keep content as long as it passes rules,
+            default:
```

Comment:
Should we really fallback to "validate" in case of an unknown action? I think we should fail if we find an unknown value. Perhaps run these actions on "validate" and ""/null?

Reply:
Not sure if my solution is better now

Reply:
https://github.com/apache/sling-org-apache-sling-xss/blob/81cc4a4f07ce77e8ac383fe3263a0787a2ffa447/src/main/java/org/apache/sling/xss/impl/CustomPolicy.java#L111-L114